File: add-binary.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        char carry = 0;
        vector<char> sum;
        int first = min(a.length(),b.length());
        int last = max(a.length(),b.length());
        for(int i=0;i<first;i++)
        {
            int x = a[a.length()-1-i]-'0',y = b[b.length()-1-i] - '0';
            sum.push_back((x^y^carry)?'1':'0');
            carry = (x&y)|(x&carry)|(carry&y);
        }
        if(a.length() == first)
        {
            for(int i=first;i<b.length();i++)
            {
                int y = b[b.length()-1-i] - '0';
                sum.push_back((y^carry)?'1':'0');
                carry &= y;
            }
        }
        if(b.length() == first)
        {
            for(int i=first;i<a.length();i++)
            {
                int x = a[a.length()-1-i] - '0';
                sum.push_back((x^carry)?'1':'0');
                carry &= x;
            }
        }
        if(carry)sum.push_back('1');
        string ans = "";
        for(int i=sum.size()-1;i>=0;i--)
            ans += sum[i];
        return ans;
    }
};

File: July LeetCoding Challenge/Add Binary.cs
public class Solution {
    
    public class Sum{
        private char carry;
        private StringBuilder sum;
        public Sum()
        {
            carry = (char)0;
            sum = new StringBuilder();
        }
        
        public void Add(char a,char b)
        {
            char s = (char)(a ^ b ^ carry);
            sum.Insert(0, (char)(s + '0'));
            carry = (char)(a & b | carry & (a ^ b));
        }
        
        public string getSum()
        {
            return carry==1? sum.Insert(0, '1').ToString(): sum.ToString();
        }
    }
    
    public string AddBinary(string a, string b) {
        Sum s = new Sum();
        int i=a.Length-1,j=b.Length-1;
        for(;i>=0&&j>=0;i--,j--)
            s.Add((char)(a[i] - '0'),(char)(b[j] - '0'));
        
        for(;i>=0;i--)
            s.Add((char)(a[i] - '0'),(char)0);
        
        for(;j>=0;j--)
            s.Add((char)0,(char)(b[j] - '0'));
        
        return s.getSum();
    }
}
